dbms/src/Encryption/KeyspacesKeyManager.cpp

```diff
@@ -0,0 +1,145 @@
+            ENCRYPTION_KEY_RESERVED_PAGEU64_ID);
+        // getFile will be called after newFile, so page must exist
+        Page page = ps_write->read(
+            page_id,
+            nullptr,
+            {},
+            /*throw_on_not_exist*/ true);
+        auto data = page.getFieldData(0);
+        return master_key->decryptEncryptionKey(String(data));
+    };
+    auto [key, exist] = keyspace_id_to_key.getOrSet<>(keyspace_id, load_func);
+    // Use MD5 of file path as IV
+    unsigned char md5_value[MD5_DIGEST_LENGTH];
+    static_assert(MD5_DIGEST_LENGTH == sizeof(uint64_t) * 2);
+    String file_path = fmt::format("{}/{}", ep.full_path, ep.file_name);
+    MD5(reinterpret_cast<const unsigned char *>(file_path.c_str()), file_path.size(), md5_value);
+    return key->generateEncryptionInfo(String(reinterpret_cast<const char *>(md5_value)));
+}
+
+FileEncryptionInfo KeyspacesKeymanager::newInfo(const EncryptionPath & ep)
+{
+    const auto keyspace_id = ep.keyspace_id;
+    if (unlikely(keyspace_id == NullspaceID)
+        || (likely(!tiflash_instance_wrap->proxy_helper->getKeyspaceEncryption(keyspace_id))))
+    {
+        return FileEncryptionInfo{
+            FileEncryptionRes::Disabled,
+            EncryptionMethod::Plaintext,
+            nullptr,
+            nullptr,
+            nullptr,
+        };
+    }
+
+    auto load_func = [this, keyspace_id]() -> EncryptionKeyPtr {
+        // Generate new encryption key
+        const auto encryption_key = master_key->generateEncryptionKey();
+        // Write encrypted key to PageStorage
+        UniversalWriteBatch wb;
+        const auto page_id = UniversalPageIdFormat::toFullPageId(
+            UniversalPageIdFormat::toLocalKVPrefix(UniversalPageIdFormat::LocalKVKeyType::EncryptionKey, keyspace_id),
+            ENCRYPTION_KEY_RESERVED_PAGEU64_ID);
+        MemoryWriteBuffer wb_buffer;
+        writeBinary(encryption_key->exportString(), wb_buffer);
+        auto read_buf = wb_buffer.tryGetReadBuffer();
+        wb.putPage(page_id, 0, read_buf, wb_buffer.count());
+        ps_write->write(std::move(wb));
+        return encryption_key;
+    };
+    auto [key, exist] = keyspace_id_to_key.getOrSet(keyspace_id, load_func);
+    // Use MD5 of file path as IV
+    unsigned char md5_value[MD5_DIGEST_LENGTH];
+    static_assert(MD5_DIGEST_LENGTH == sizeof(uint64_t) * 2);
+    String file_path = fmt::format("{}/{}", ep.full_path, ep.file_name);
+    MD5(reinterpret_cast<const unsigned char *>(file_path.c_str()), file_path.size(), md5_value);
+    return key->generateEncryptionInfo(String(reinterpret_cast<const char *>(md5_value)));
+}
+
+void KeyspacesKeymanager::deleteInfo(const EncryptionPath & ep, bool /*throw_on_error*/)
+{
+    const auto keyspace_id = ep.keyspace_id;
+    if (unlikely(keyspace_id == NullspaceID)
+        || (likely(!tiflash_instance_wrap->proxy_helper->getKeyspaceEncryption(keyspace_id))))
+        return;
+    // do nothing
```

Comment:
Seems this function ends up doing nothing?

Reply:
yes, I change it to a LRU cache, so there is no ne need to remove here.